articles/clojure_elixir_fibonacci_mstalker/article.md

```diff
@@ -0,0 +1,339 @@
+
+Here's today's programming challenge:
+
+> For any non-negative integer `n`, write a function to retrieve the `n`th Fibonacci number. Use a zero-based index.
+
+For example, when `n` is `4`, our function should return `3`.
+
+My goal in the following code exploration is to highlight some language features of Clojure and Elixir. I'll give examples of how someone might solve our programming challenge in either language, and then I'll explain some syntax or conventions. I'm not attempting to produce the most efficient or the most clever code, as that may actually get in the way of giving you a taste for each language!
+
+
+### Clojure
+
+A straightforward solution to our problem looks like this in Clojure:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (= n 0) 0
+    (= n 1) 1
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We created a `fibonacci` namespace on the first line, and then we used `defn` to define a `compute` function. The next line, `[n]`, indicates that the function has one argument, which we call `n`.
+
+You'll notice all the parentheses. A pair of parentheses creates a list. The first element of a list is generally a function or macro, and the rest of the elements are the arguments.
+
+`cond` begins a conditional. If `n` is equal to zero, the function returns `0`. If `n` is `1`, then `compute` returns `1`. Mathematical operators like `=`, `+`, and `-` are just functions in Clojure, and they follow the same syntactical rules. That's why these operators appear before their arguments, rather than between them.
+
+You might have observed that no commas separate function arguments. That's normal in Clojure. Commas are optional in collections.
+
+Finally, we use the `:else` keyword to do something if `n` isn't `0` or `1`. We recursively call the `compute` function twice, passing in different arguments. This sums the two previous Fibonacci values. Putting the two `compute` function calls on separate lines is aesthetic. I could have written this:
+
+```clojure
+:else (+ (compute (- n 1)) (compute (- n 2)))
+```
+
+Clojure has no explicit `return`, so functions just return the last value they evaluate.
+
+If we wanted to, we could shorten the implementation slightly by using an inequality:
+
+```clojure
+(ns fibonacci)
+
+(defn compute
+  [n]
+  (cond
+    (< n 2) n
+    :else (+ (compute (- n 1))
+             (compute (- n 2)))))
+```
+
+We can call our function like this:
+
+```clojure
+(fibonacci/compute 10)
+;; 55
+```
+
+A semicolon starts a comment in Clojure. I wrote the return value of our function call after `;;`. Using two semicolons is idiomatic, and the language would have allowed us to use just one.
+
+
+### Elixir
+
+How does Elixir handle the simple implementation? Here is how we could rewrite the Clojure code in Elixir:
+
+```elixir
+defmodule Fibonacci do
+  def compute(n) do
+    cond do
+      n == 0 -> 0
```

Comment:
If you're going to compare the examples I think it's worth making the code equivalent:

```elixir
defmodule Fibonacci do
  def compute(n) do
    cond do
      n < 2 -> n
      true -> compute(n - 1) + compute(n - 2)
    end
  end
end
```

Reply:
I see what you mean. My goal was to write an implementation in each language using both 0 and 1 explicitly, and then to write an implementation where I consolidate the 0 and 1 cases using an inequality or a guard clause. Maybe I'll just take the inequality and guard clause cases out. It doesn't really add anything to the article to include the Clojure inequality, and omitting the Elixir `n in [0, 1]` won't hurt, either. I use a guard clause later in the article. What do you think?

Reply:
If I remove some code samples, do you think it would be better to eliminate this version:
```clojure
(defn compute
  [n]
  (cond
    (= n 0) 0
    (= n 1) 1
    :else (+ (compute (- n 1))
             (compute (- n 2)))))
```
or this version?
```clojure
(defn compute
  [n]
  (cond
    (< n 2) n
    :else (+ (compute (- n 1))
             (compute (- n 2)))))
```